crates/runtime/src/memfd.rs

```diff
@@ -0,0 +1,566 @@
+        // memory filled with zeroes.
+        if let Some(image) = maybe_image {
+            if image.len > 0 {
+                let image = image.clone();
+
+                unsafe {
+                    let ptr = rustix::io::mmap(
+                        (self.base + image.offset) as *mut c_void,
+                        image.len,
+                        rustix::io::ProtFlags::READ | rustix::io::ProtFlags::WRITE,
+                        rustix::io::MapFlags::PRIVATE | rustix::io::MapFlags::FIXED,
+                        image.fd.as_file(),
+                        image.offset as u64,
+                    )
+                    .map_err(|e| InstantiationError::Resource(e.into()))?;
+                    assert_eq!(ptr as usize, self.base + image.offset);
+                }
+
+                self.image = Some(image);
+            }
+        }
+
+        // mprotect the initial `initial_size_bytes` to be accessible.
+        self.initial_size = initial_size_bytes;
+        self.cur_size = initial_size_bytes;
+        self.set_protection(
+            0,
+            initial_size_bytes,
+            rustix::io::MprotectFlags::READ | rustix::io::MprotectFlags::WRITE,
+        )
+        .map_err(|e| InstantiationError::Resource(e.into()))?;
+
+        self.dirty = true;
+        Ok(())
+    }
+
+    pub(crate) fn clear_and_remain_ready(&mut self) -> Result<()> {
+        assert!(self.dirty);
+        // madvise the image range. This will throw away dirty pages,
+        // which are CoW-private pages on top of the initial heap
+        // image memfd.
+        unsafe {
+            rustix::io::madvise(
+                self.base as *mut c_void,
+                self.static_size,
+                rustix::io::Advice::LinuxDontNeed,
+            )?;
+        }
+
+        // mprotect the initial heap region beyond the initial heap size back to PROT_NONE.
+        self.set_protection(
+            self.initial_size,
+            self.static_size - self.initial_size,
+            rustix::io::MprotectFlags::empty(),
+        )?;
+        self.dirty = false;
+        Ok(())
+    }
+
+    fn set_protection(
+        &self,
+        start: usize,
+        len: usize,
+        flags: rustix::io::MprotectFlags,
+    ) -> Result<()> {
+        assert!(start.checked_add(len).unwrap() <= self.static_size);
+        let mprotect_start = self.base.checked_add(start).unwrap();
+        if len > 0 {
+            unsafe {
+                rustix::io::mprotect(mprotect_start as *mut _, len, flags)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn has_image(&self) -> bool {
+        self.image.is_some()
+    }
+
+    pub(crate) fn is_dirty(&self) -> bool {
+        self.dirty
+    }
+
+    /// Map anonymous zeroed memory across the whole slot, with the
+    /// given protections. Used both during instantiate and during
+    /// drop.
+    fn map_anon_memory(&self, prot: rustix::io::ProtFlags) -> Result<()> {
```

Comment:
I think now the `prot` flag is always `ProtFlags::empty()` so this could perhaps be renamed to `reset_with_anon_memory` or something like that? (with no arguments)

Reply:
Done!